swap/swap.go

```diff
@@ -1,98 +1,636 @@
+	cheque := msg.Cheque
+	log.Info("received cheque from peer", "peer", p.ID().String())
+	actualAmount, err := s.processAndVerifyCheque(cheque, p)
+	if err != nil {
+		return err
+	}
+
+	log.Debug("received cheque processed and verified", "peer", p.ID().String())
+
+	// reset balance by amount
+	// as this is done by the creditor, receiving the cheque, the amount should be negative,
+	// so that updateBalance will calculate balance + amount which result in reducing the peer's balance
+	s.accountingLock.Lock()
+	err = s.resetBalance(p.ID(), 0-int64(cheque.Honey))
+	s.accountingLock.Unlock()
+	if err != nil {
+		return err
+	}
+
+	// cash in cheque
+	opts := bind.NewKeyedTransactor(s.owner.privateKey)
+	opts.Context = ctx
+
+	otherSwap, err := contract.InstanceAt(cheque.Contract, s.backend)
+	if err != nil {
+		return err
+	}
+
+	// submit cheque to the blockchain and cashes it directly
+	go func() {
+		// blocks here, as we are waiting for the transaction to be mined
+		receipt, err := otherSwap.SubmitChequeBeneficiary(opts, s.backend, big.NewInt(int64(cheque.Serial)), big.NewInt(int64(cheque.Amount)), big.NewInt(int64(cheque.Timeout)), cheque.Signature)
+		if err != nil {
+			// TODO: do something with the error
+			// and we actually need to log this error as we are in an async routine; nobody is handling this error for now
+			log.Error("error submitting cheque", "err", err)
+			return
+		}
+		log.Debug("submit tx mined", "receipt", receipt)
+
+		receipt, err = otherSwap.CashChequeBeneficiary(opts, s.backend, s.owner.Contract, big.NewInt(int64(actualAmount)))
+		if err != nil {
+			// TODO: do something with the error
+			// and we actually need to log this error as we are in an async routine; nobody is handling this error for now
+			log.Error("error cashing cheque", "err", err)
+			return
+		}
+		log.Info("Cheque successfully submitted and cashed")
+	}()
+	return err
+}
+
+// processAndVerifyCheque verifies the cheque and compares it with the last received cheque
+// if the cheque is valid it will also be saved as the new last cheque
+func (s *Swap) processAndVerifyCheque(cheque *Cheque, p *Peer) (uint64, error) {
+	if err := s.verifyChequeProperties(cheque, p); err != nil {
+		return 0, err
+	}
+
+	lastCheque := s.loadLastReceivedCheque(p)
+
+	// TODO: there should probably be a lock here?
+	expectedAmount, err := s.oracle.GetPrice(cheque.Honey)
+	if err != nil {
+		return 0, err
+	}
+
+	actualAmount, err := verifyChequeAgainstLast(cheque, lastCheque, expectedAmount)
+	if err != nil {
+		return 0, err
+	}
+
+	if err := s.saveLastReceivedCheque(p, cheque); err != nil {
+		log.Error("error while saving last received cheque", "peer", p.ID().String(), "err", err.Error())
+		// TODO: what do we do here? Related issue: https://github.com/ethersphere/swarm/issues/1515
+	}
+
+	return actualAmount, nil
+}
+
+// verifyChequeProperties verifies the signature and if the cheque fields are appropriate for this peer
```

Comment:
I changed it so that all of them are defined on cheque now. I wouldn't combine them as having them separate makes testing them easier. Anyway the verify functions probably will become quite a bit simpler once we switch over to the waiver-free `SimpleSwap` (which is ready from the contract side, just waiting for this PR to be merged) then we should reconsider combining them.